.github/workflows/tests.yml

```diff
@@ -64,71 +63,137 @@ jobs:
+
+            const runners = ["11-arm64", "11", "10.15", "10.14"]
+
+            if (label_names.includes('CI-linux-self-hosted')) {
+              runners.push('linux-self-hosted-1')
+            } else {
+              runners.push('ubuntu-latest')
             }
 
+            core.setOutput('runners', JSON.stringify(runners))
+
   tests:
-    needs: tap_syntax
-    if: github.event_name == 'pull_request' && needs.tap_syntax.outputs.syntax-only == 'false'
+    needs: pre_tests
+    if: github.event_name == 'pull_request' && needs.pre_tests.outputs.syntax-only == 'false'
     strategy:
       matrix:
-        version: ["11-arm64", "11", "10.15", "10.14"]
+        runner: ${{fromJson(needs.pre_tests.outputs.runners)}}
       fail-fast: false
-    runs-on: ${{ matrix.version }}
+    runs-on: ${{matrix.runner}}
     timeout-minutes: 4320
     env:
-      PATH: "/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin"
-      HOMEBREW_DEVELOPER: 1
-      GITHUB_ACTIONS_HOMEBREW_SELF_HOSTED: 1
       GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
       HOMEBREW_GITHUB_API_TOKEN: ${{secrets.GITHUB_TOKEN}}
     steps:
+      - name: Set environment variables
+        run: |
+          if [ "$RUNNER_OS" = 'macOS' ]; then
+            echo 'PATH=/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin' >> $GITHUB_ENV
+            echo 'GITHUB_ACTIONS_HOMEBREW_SELF_HOSTED=1' >> $GITHUB_ENV
+          else
+            echo 'HOMEBREW_CORE_GIT_REMOTE=${{github.event.repository.html_url}}' >> $GITHUB_ENV
+            echo 'HOMEBREW_FORCE_HOMEBREW_ON_LINUX=1' >> $GITHUB_ENV
+          fi
+
       - name: Check whether ARM tests will be forced
-        if: needs.tap_syntax.outputs.force-arm == 'false'
+        if: runner.os == 'macOS' && needs.pre_tests.outputs.force-arm == 'false'
         run: echo 'HOMEBREW_SKIP_UNBOTTLED_ARM_TESTS=1' >> $GITHUB_ENV
 
+      - name: Check whether Linux tests will be forced
+        if: runner.os == 'Linux' && needs.pre_tests.outputs.force-linux == 'false'
+        run: echo 'HOMEBREW_SKIP_UNBOTTLED_LINUX_TESTS=1' >> $GITHUB_ENV
+
       - name: Set up Homebrew
         id: set-up-homebrew
+        if: runner.os == 'macOS'
         uses: Homebrew/actions/setup-homebrew@master
 
-      - run: brew test-bot --only-cleanup-before
+      - if: runner.os == 'macOS'
+        run: brew test-bot --only-cleanup-before
+
+      - if: runner.os == 'macOS'
+        run: brew test-bot --only-setup
 
-      - run: brew test-bot --only-setup
+      - name: Run Docker container
+        if: runner.os == 'Linux'
+        run: |
+          docker run \
+            --detach \
+            --user linuxbrew \
+            --name linuxbrew \
+            --env-file <(env | egrep 'HOMEBREW|GITHUB') \
+            --workdir /tmp/bottles \
+            ghcr.io/homebrew/ubuntu16.04:master \
+            sleep inf
+          # Fix working directory permissions
+          docker exec --user root linuxbrew chmod 777 /tmp/bottles
```

Comment:
Isn't this something we can do with the actions Docker DSLs? If not, it'd be nice to pull this into `Homebrew/actions/setup-homebrew` with a `docker` arg or something.

Reply:
Can't specify `container` at the job level conditionally. Cause for macOS the value for this key would be empty and that is not supported as of now.

Reply:
Ah, shame, no worries, thanks for answering.



> it'd be nice to pull this into `Homebrew/actions/setup-homebrew` with a `docker` arg or something.

Thoughts on this?

Reply:
Yea, thought about it earlier too. Those label checks step could be pulled out too.

Reply:
Yeh, that'd be nice I think ❤️